Install plugin

binocarlos opened the pull request:
This PR is a `plugin.py` script that loops over each of the nodes in `c.config["agent_nodes"]` and for each of them:
- generate a `plugin.crt` and `plugin.key` and upload them to `/etc/flocker` on the node
- git clone the flocker-docker-plugin and checkout a branch
- install pip and then pip install flocker-docker-plugin
- write a service file pointing to a bash script wrapper - ubuntu & centos
- stop docker - download docker binary - start docker

There are environment variables for controlling where to get docker from and where to clone the plugin from.

I pointed the default docker binary at one I built from docker master earlier (so it should have `--volume-driver`)

The default docker-plugin repo is a [branch I made](https://github.com/ClusterHQ/flocker-docker-plugin/tree/txflocker-env-vars) in the clusterhq repo - it means I can set the filenames of the certs to `plugin.{crt,key}`

binocarlos committed:
the start of the script to upload and configure the plugin alongside TLS certs

binocarlos committed:
generate a plugin user cert for each node - use the IP address of the node to name the certs so they are not overwritten

binocarlos committed:
once the plugin certs are generated - upload them into /etc/flocker/plugin.crt /etc/flocker/plugin.key

binocarlos committed:
comments for the cert generation step

binocarlos committed:
configure the location of the docker binary

we download a new docker binary and install it on the users system - this variable controls where the binary is downloaded from

binocarlos committed:
allow the name of the docker service to be configured

clusterhq-flocker-node uses the docker.io service name - we need to stop and start the docker service in order to replace the binary

binocarlos committed:
for each node - stop the docker service, download the latest docker binary to /usr/bin/docker and then start the docker service

binocarlos committed:
needed to import os

binocarlos committed:
speller - also break the cert loop for testing

binocarlos committed:
consistent tabbing

binocarlos committed:
refactor the settings section so its a loop over a default dict

binocarlos committed:
notes about the provenence of the docker binary hosted on google storage

binocarlos committed:
note about why we need the servicename

binocarlos committed:
first pass at configuring the flocker-plugin on the nodes

binocarlos committed:
careful usage of runSSHRaw, installing pip - then pip install

also populate env variables for plugin.crt and plugin.key

binocarlos committed:
point to the clusterhq repo and vary the branch

binocarlos committed:
notes about the plugin.py in the Readme - include a list of the environment variables that control it

binocarlos committed:
speller

binocarlos committed:
make it clear that the certs are uploaded into the /etc/flocker folder

binocarlos committed:
make plugin.py +x

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+  # allow env override for where to download the experimental docker binary from
+  # the docker-volumes binary is a buid from latest docker/master: 
+  # s4caa939 - https://github.com/docker/docker/tree/4caa9392f8aa4e57bfe43880b5f67d15b00ed8a7
+  'DOCKER_BINARY_URL':'http://storage.googleapis.com/experiments-clusterhq/docker-binaries/docker-volumes',
```
Add space after `:` in all these, line-wrap all lines to 76 chars

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+settings = {}
+
+# loop over each of the default vars and check to see if we have been given an override in the environment
+for field in settings_defaults:
```
4 spaces for indent, not 2

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+    for node in c.config["agent_nodes"]:
+        
+        # use the node IP to name the local files so they do not overwrite each other
+        c.run("flocker-ca create-api-certificate %s" % (node + '-plugin',))
```
`"flocker-ca create-api-certificate %s-plugin" % (node,)` would be nicer (avoid mixing interpolation and concatenation)

binocarlos replied:
wow yes - not sure what I was thinking there

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+        plugin_repo_folder = settings['PLUGIN_REPO'].split('/').pop()
+
+        # the full api path to the control service
+        controlservice = 'https://%s:4523/v1' % (control_ip)
```
comma after `control_ip` -- string interpolation should always take a one-tuple, and `(x,)` is how to spell a 1-tuple (otherwise if `control_ip` happens to be a 2-tuple in the future, it would blow up mysteriously)

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+        # pip install the plugin
+        c.a(node, "pip install -r /root/%s/requirements.txt" % (plugin_repo_folder))
+        
+        print "Have control service: %s" % (controlservice)
```
,

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+            c.runSSHRaw(node, "yum install -y python-devel python-pip")
+
+        # pip install the plugin
+        c.a(node, "pip install -r /root/%s/requirements.txt" % (plugin_repo_folder))
```
,

lukemarsden replied:
what is `c.a`?

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+respawn
+
+# Start the process
+exec /usr/bin/env bash /root/runflockerplugin.sh
```
I'd rather that this just ran the flocker plugin. Can you move the env var configuration into the upstart script?

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+Description=flocker-plugin - flocker-plugin job file
+
+[Service]
+ExecStart=/usr/bin/env bash /root/runflockerplugin.sh
```
ditto

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+    # loop each agent and get the plugin installed/running
+    # clone the plugin and configure an upstart/systemd unit for it to run
+    for node in c.config["agent_nodes"]:
+
```
style - remove newlines after `for` loops

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+
+    print "Replacing docker binary"
+    # download and replace the docker binary on each of the nodes
+    for node in c.config["agent_nodes"]:
```
style - remove newlines after `for` loops

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+        if c.config["os"] == "ubuntu":
+          c.runSSHRaw(node, "stop %s" % (settings['DOCKER_SERVICE_NAME']))
+        elif c.config["os"] == "centos":
+          c.runSSHRaw(node, "systemctl stop %s.service" % (settings['DOCKER_SERVICE_NAME']))
```
,

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+
+        # download the latest docker binary\
+        print "Downloading the latest docker binary on %s - %s" % (node, settings['DOCKER_BINARY_URL'])
+        c.runSSHRaw(node, "wget -O /usr/bin/docker %s" % (settings['DOCKER_BINARY_URL']))
```
,

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+        # stop the docker service
+        print "Starting the docker service on %s" % (node)
+        if c.config["os"] == "ubuntu":
+          c.runSSHRaw(node, "start %s" % (settings['DOCKER_SERVICE_NAME']))
```
4 spaces not 2

lukemarsden replied:
,

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+        if c.config["os"] == "ubuntu":
+          c.runSSHRaw(node, "start %s" % (settings['DOCKER_SERVICE_NAME']))
+        elif c.config["os"] == "centos":
+          c.runSSHRaw(node, "systemctl start %s.service" % (settings['DOCKER_SERVICE_NAME']))
```
,

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+        elif c.config["os"] == "centos":
+          c.runSSHRaw(node, "systemctl start %s.service" % (settings['DOCKER_SERVICE_NAME']))
+
+
```
remove trailing newlines

lukemarsden commented:
Thanks for this, looks like very useful functionality! Please address the review comments and resubmit for review.

lukemarsden commented on plugin.py:
```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env bash
+rm -f /usr/share/docker/plugins/flocker.sock || true
+export FLOCKER_CONTROL_SERVICE_BASE_URL=%s
+export MY_NETWORK_IDENTITY=%s
```
as we discovered, this is sometimes (AWS) the wrong IP!

binocarlos committed:
keep line lengths down to less than 79 for PEP 8

binocarlos committed:
add space after settings_defaults keys

binocarlos committed:
use 4 spaces for indentation

Commit:
add support for private ips in a breaking change to cluster.yml format

binocarlos committed:
dont mix concatenation and interpolation

binocarlos committed:
trailing commas for tuple interpolation values

binocarlos committed:
c.a isn't a thing - it should be SSHing

binocarlos committed:
remove blank lines after for loops

binocarlos committed:
ensure 4 spaces in tabs

binocarlos committed:
remove trailing newlines

binocarlos committed:
for the upstart/systemd jobs - invoke twistd directly - loose the bash script approach

binocarlos committed:
s/public_ip/public/

binocarlos committed:
use the public ip for SSHing into the node and the private ip for the nodes identity

binocarlos committed:
yikes missed one s/node/public_ip/

binocarlos committed:
if docker is not running dont fail trying to stop it

binocarlos committed:
line length < 76

binocarlos committed:
ensure that /usr/share/${DOCKER_SERVICE_NAME}/plugins exists as a folder before trying to start the plugin

binocarlos committed:
syntax error with missing bracket

binocarlos committed:
configure the DOCKER_PLUGINS_FOLDER for the plugin so we are writing flocker.sock to /usr/share/docker.io/plugins - this is configured based on the DOCKER_SERVICE_NAME var

binocarlos committed:
the plugins folder is fixed to /usr/share/docker/plugins

binocarlos committed:
dups missing bracket

binocarlos committed:
dont remove the plugin repo and fail if git clone fails

binocarlos committed:
create /usr/share/docker/plugins for both Arches

binocarlos committed:
docker needs replacing and restarting before the plugin starts because docker restarting removes /usr/share/docker/plugins

binocarlos committed:
have option to skip downloading docker and downloading / installing plugin

    this is useful for cases where the user has already baked these things into images

lukemarsden commented:
looks good, merging!
